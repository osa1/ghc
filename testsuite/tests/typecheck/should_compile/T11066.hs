{-# LANGUAGE GADTs #-}

module T11066 where

data Foo a where
  Foo1 :: Foo Char
  Foo2 :: Foo Int

data TyEquality a b where
  Refl :: TyEquality a a

checkTEQ :: Foo t -> Foo u -> Maybe (TyEquality t u)
checkTEQ x y = error "unimportant"

-- Problem: Operationally, step1 and step2 should be the same. But previously
-- the type checker was rejecting step2 and accepting step1. See also #11066.

step1 :: Bool
step1 =
 let foo d =
       case checkTEQ Foo1 d of
         Just Refl -> True
         Nothing -> False
 in foo Foo2

step2 :: Bool
step2 = case checkTEQ Foo1 Foo2 of
          Just Refl -> True -- Previously an "inaccessible code" error.
          Nothing -> False
