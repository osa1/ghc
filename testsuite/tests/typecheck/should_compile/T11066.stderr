
T11066.hs:25:16: warning:
    Couldn't match type ‘Char’ with ‘Int’
    Inaccessible code in
      a pattern with constructor: Refl :: forall a. TyEquality a a,
      in a case alternative
    In the pattern: Refl
    In the pattern: Just Refl
    In a case alternative: Just Refl -> True
